### PR TITLE
Hover Info improvements

### PR DESCRIPTION
**Pull request by ChrisPenner**
## Overview

Massively improves reliability and correctness of hover info; while also adding a lot of combinators which will make future features easier to implement. Specifically, this adds combinators for identifying which ABT term a given file position relates to, and what references that node contains if any.

The previous implementation simply searched the codebase for the name under the cursor, which meant it couldn't do any TDNR, and there was no guarantee that what it found would be unique, or would match what the parser did.

Now I use the actual parser output; so it's guaranteed to match. This required a fair amount of work to translate from a file position into the actual reference from the parser, but now that it's done it unlocks a bunch of other LSP features, like Jump-To-Definition!

I also added syntax highlighting to the hover help.

Here's how some things have changed:

Old Term hover:

image

New Term hover:

image

---

Old Type sig hover:

image

New Type sig hover:

image

---

Old ability constructor hover:

image

New ability constructor hover:

image

---

Old type ref in constructor hover ( No hover ):

image

New type ref in constructor hover (shows decl):

image

---

Old literal hover (doesn't understand string literals):

image

New literal hover (Shows literal type):

image

## Implementation notes

* Adds a FileSummary to the LSP state which collects as much information as possible from the parsed and typechecked files so that we don't need to implement each endpoint twice for it to work when typechecking and not typechecking.
* Add queries which when given a position will crawl into terms and find the smallest term node which still contains that position
* Use the above queries to use the typechecked/parsed files for resolving the Ref under the cursor
* Look up that Ref in either the file or the codebase and render its type info
* Wrap Hover results in a markdown block so the LSP will highlight it.
* Changes monoid for Ann to return the range that includes both provided Anns, rather than return invalid Anns where the start is after the end.

## Interesting/controversial decisions

* For the literals I just hard-coded the literal type names, there could be a better way to do this, but I think this is more than good enough for now.

## Test coverage

I've tested all the cases I provided screenshots for, I'd like to figure out some form of automated testing for LSP stuff at some point, but haven't figured out a great solution for that yet.

## Loose ends

* Hover info should include docs, I'll add that in a separate PR
* We can provide range info for on-hover tool-tips which VS code can use for cool code-peeking features, but unfortunately it would only work for code that's being edited in the file, and it's a fair bit of work to do right now, it'll be easier to add after Jump-to-definition is implemented.

### CODE REVIEW

File: unison-cli/src/Unison/LSP/Types.hs [hunk 1]
```diff
@@ -105,14 +112,31 @@ data FileAnalysis = FileAnalysis
...
   }
   deriving (Show)
 
+-- | A file that parses might not always type-check, but often we just want to get as much
+-- information as we have available. This provides a type where we can summarize the
+-- information available in a Unison file.
+--
+-- If the file typechecked then all the Ref Ids and types will be filled in, otherwise
+-- they will be Nothing.
+data FileSummary = FileSummary
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/LSP/Types.hs]: As the doc says, this is a really helpful type for allowing the LSP endpoint implementations to not really need to care whether the current file type-checks or not, they can just use as much information as is available without needing to implement the endpoint twice.

File: unison-syntax/src/Unison/Parser/Ann.hs [hunk 2]
```diff
@@ -20,9 +20,28 @@ instance Monoid Ann where
   mempty = External
 
 instance Semigroup Ann where
-  Ann s1 _ <> Ann _ e2 = Ann s1 e2
+  Ann s1 e1 <> Ann s2 e2 = Ann (min s1 s2) (max e1 e2)
```
Comment (ChrisPenner) [inline unison-syntax/src/Unison/Parser/Ann.hs]: When combining Ann's we want the full span of those Anns.

The old monoid was giving invalid Anns for things like Type arrows where the start was after the end, etc.

**Comment by ceedubs** [comment]
🎉 woohoo nice @ChrisPenner this is super exciting!

I noticed kind of a weird issue.

If I have `main = do Either.toException (Right 3)` and hover on `toException` then it says no hover found. Strangely if I change it to `main = do Either.toException 3` then the hover actually _does_ give the correct result (and additionally I correctly get the compile error message).

There are a couple of other terms named `toException` in base. I'm not sure whether that's part of the issue.

**Comment by ChrisPenner** [comment]
@ceedubs  cool, thanks for the report!
I think this may be part of the weird Annotations issues I've noticed with do-blocks (which I plan to address in a separate PR since they can get a bit hairy), but I'll try to confirm that that's the issue here as well before merging 😄 

You'll also notice it stops working properly if you've got multiple lines in a `do` block 🙃  (because of the same annotation bugs in the parser)

**Comment by ChrisPenner** [comment]
Huh, looks like something in either the synthesis process or TDNR is clobbering/changing annotations, so ironically you get better hover results if your file doesn't typecheck haha.

I'll dig deeper into how TDNR is messing with annotations 👍🏼

**Comment by ChrisPenner** [comment]
Making some notes on broken annotations that I find here:

1. `effect` and `effect1` in `Unison.Type` erroneously tag their effects list with the annotation for the remainder of the arrow, this results in false matches when hovering on type signatures

E.g. Hovering over `Nat` in the following will match a hidden and empty effects list which contains the right -side of the arrow.

```unison
thingy : Text -> Nat
thingy  = do 1 + 1
```

2. Top level bindings with args left of the `=` seem to shadow the top-level type annotation underneath a lambda somehow, e.g. hover over 'Nat' here doesn't work, since it somehow matches the lambda for `User _ -> ...`;
```unison
thingy :  Text -> Nat
thingy _ = 1
```

3. Annotations aren't combined correctly in do-blocks/let-blocks

E.g. you can't hover anything other than the first line:

```unison
thingy : Nat
thingy = let
  1 + 1
  23
  4556
```

4. Binary expressions don't contain the annotation for their left-hand-side; 
e.g. you can hover over `false` but not `true`

```unison
thingy : Boolean
thingy = true && false
```

I'll keep track of those here: https://github.com/unisonweb/unison/issues/3639

**Review by pchiusano — APPROVED** [review]
Looks sweet!

**Comment by pchiusano** [comment]
Weird, it looks like https://github.com/unisonweb/unison/pull/3751 was set to auto-close this when merged.

**Comment by ChrisPenner** [comment]
Yeah, I think I accidentally said that it depends on `these fixes ...` and Github thought it solved this pr or something, no worries.

**Comment by ChrisPenner** [comment]
Made a little progress on this today, there are still some issues, namely that the typechecker will copy the in-file annotations onto the types it infers, it's great to have those annotations on the "type annotation node", but no bueno to have annotations on the actual type nodes, because those imply the type definition is in the file when it's not.

I think I've got a work-around for that in the meantime but will need to do more testing to see how it feels.

- [x] Also, destructuring binds are still broken, I know how to fix that, just haven't gotten to it yet.
- [x] Also, annotations on effect blocks are also broken, so hover doesn't work on anything past the first argument in types with arrows 🤦🏼‍♂️

**Comment by ChrisPenner** [comment]
Okay, I think this is finally ready for another try!

I've sorted out the problems I knew about 😄. 

It still has the following limitations:

* Can't get hovers for local variables (a restriction of how we do typechecking right now)
* Can't get hovers for symbols in the same file (I can add this, but will do that in a separate PR)

But both of those limitations exist for the current hover anyways, so we're not losing anything.

There's one remaining glitch where if you hover over the `,` on a tuple in a destructuring bind it might give you the wrong name in the tooltip, but I think that's both rare and minor enough that this is worth merging as-is; it's still a good step above what we have right now.


Aside from one new combinator and some new exports/instances this doesn't make any changes to things outside of the LSP, so unless you're interested in the implementation probably doesn't need much review, it's pretty unlikely it could break anything in the core experience 🤷🏼‍♂️

File: unison-core/src/Unison/Pattern.hs [hunk 3]
```diff
@@ -37,6 +37,23 @@ data Pattern loc
   | SequenceOp loc (Pattern loc) !SeqOp (Pattern loc)
   deriving (Ord, Generic, Functor, Foldable, Traversable)
 
+annotation :: Pattern loc -> loc
```
Comment (pchiusano) [inline unison-core/src/Unison/Pattern.hs]: You could provide an instance of the `Annotated` typeclass.

**Review by pchiusano — APPROVED** [review]
Cool! I think any of the LSP stuff is fine to have a light review, okay to merge as is, and you should just let us know if there's something you want someone else to take a detailed look at. Left one minor comment for you.

**Review by ceedubs — APPROVED** [review]
🎉 I gave this a spin for ~5 minutes and it's been working great so far!

**Pull request merged**